### PR TITLE
Snapcast

### PR DESCRIPTION
**Pull request by awiouy**
This adds snapclient as a service addon.
I will add a corresponding snapserver addon.

### CODE REVIEW

File: packages/addons/addon-depends/snapcast-depends/snapcast/package.mk [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+PKG_SITE="https://github.com/badaix/snapcast"
+PKG_URL="https://github.com/badaix/$PKG_NAME/archive/$PKG_VERSION.tar.gz"
+PKG_DEPENDS_TARGET="toolchain aixlog alsa-lib asio jsonrpcpp avahi flac libvorbis popl"
+PKG_SECTION="tools"
+PKG_SHORTDESC="snapcast"
+PKG_LONGDESC="Synchronous multi-room audio player"
+
+PKG_TOOLCHAIN="make"
+
+pre_configure_target() {
```
Comment (CvH) [inline packages/addons/addon-depends/snapcast-depends/snapcast/package.mk]: shouldn't it be ?
```
pre_configure_target() {
  rm -rf externals
  mkdir externals
    cp -r $(get_build_dir aixlog) externals/aixlog
    cp -r $(get_build_dir asio) externals/asio
    cp -r $(get_build_dir jsonrpcpp) externals/jsonrpcpp
    cp -r $(get_build_dir popl) externals/popl
    cp externals/jsonrpcpp/lib/json.hpp externals/json.hpp
}
```

File: packages/addons/service/snapclient/package.mk [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+PKG_NAME="snapclient"
+PKG_VERSION="fcf5997"
+PKG_REV="99h"
+PKG_ARCH="any"
+PKG_LICENSE="GPLv3"
+PKG_DEPENDS_TARGET="toolchain snapcast"
+PKG_SECTION="service"
+PKG_SHORTDESC="Snapclient: Synchronous multi-room audio client"
+PKG_LONGDESC="Snapclient ($PKG_VERSION) is a Snapcast client. Snapcast is a multi-room client-server audio system, where all clients are time synchronized with the server to play perfectly synced audioplays."
+PKG_AUTORECONF="no"
```
Comment (CvH) [inline packages/addons/service/snapclient/package.mk]: is no more, use `PKG_TOOLCHAIN`

File: packages/addons/service/snapclient/package.mk [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+PKG_IS_ADDON="yes"
+PKG_ADDON_NAME="Snapclient"
+PKG_ADDON_TYPE="xbmc.service.library"
+PKG_MAINTAINER="Anton Voyl (awiouy)"
+
+PKG_TOOLCHAIN="manual"
```
Comment (CvH) [inline packages/addons/service/snapclient/package.mk]: rm and replace `PKG_AUTORECONF="no"` ^^^^

File: packages/addons/service/snapclient/source/default.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+################################################################################
+#      This file is part of LibreELEC - https://libreelec.tv
+#      Copyright (C) 2018 Team LibreELEC
```
Comment (CvH) [inline packages/addons/service/snapclient/source/default.py]: 2018-present

File: packages/addons/service/snapclient/source/resources/language/English/strings.po [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Kodi Media Center language file
```
Comment (CvH) [inline packages/addons/service/snapclient/source/resources/language/English/strings.po]:
```
# Kodi Media Center language file
# Addon Name: snapclient
# Addon id: service.snapclient
# Addon Provider: Team LibreELEC
msgid ""
msgstr ""
```
to follow the standard (what Kodi is doing)

File: packages/addons/service/snapclient/source/resources/language/English/strings.po [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+msgctxt "#30010"
+msgid "Manage onboard audio"
+msgstr ""
+
+msgctxt "#30011"
+msgid "Onboard audio playback route"
+msgstr ""
+
+msgctxt "#30012"
+msgid "auto detect"
```
Comment (CvH) [inline packages/addons/service/snapclient/source/resources/language/English/strings.po]: `msgid "Autodetect"` ? (idk where it is shown)

File: packages/addons/service/snapclient/source/resources/language/English/strings.po [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+msgstr ""
+
+msgctxt "#30013"
+msgid "headphone jack"
```
Comment (CvH) [inline packages/addons/service/snapclient/source/resources/language/English/strings.po]: `msgid "Headphone Jack"` ? (idk where it is shown)

**Comment by awiouy** [comment]
Ouch @CvH :)
Will fixup :))

File: packages/addons/service/snapclient/package.mk (hunk 3 above)
Comment (awiouy) [reply packages/addons/service/snapclient/package.mk]: Hm...
Is that not the purpose of PKG_TOOLCHAIN="manual"? No autoreconf/configure_target/make_target/makeinstall_target...

File: packages/addons/service/snapclient/source/resources/language/English/strings.po (hunk 6 above)
Comment (awiouy) [reply packages/addons/service/snapclient/source/resources/language/English/strings.po]: Same as in Librespot

File: packages/addons/service/snapclient/source/resources/language/English/strings.po (hunk 7 above)
Comment (awiouy) [reply packages/addons/service/snapclient/source/resources/language/English/strings.po]: Same as in Librespot

File: packages/addons/service/snapclient/package.mk (hunk 3 above)
Comment (CvH) [reply packages/addons/service/snapclient/package.mk]: you had `PKG_AUTORECONF="no"` at the position where this `PKG_TOOLCHAIN="manual"` should have been :) so just replace `PKG_AUTORECONF="no"` -> `PKG_TOOLCHAIN="manual"`

File: packages/addons/addon-depends/snapcast-depends/snapcast/package.mk (hunk 1 above)
Comment (awiouy) [reply packages/addons/addon-depends/snapcast-depends/snapcast/package.mk]: Is that 'directory' indentatition systematic in LibreELEC?

File: packages/addons/service/snapclient/source/resources/language/English/strings.po (hunk 6 above)
Comment (CvH) [reply packages/addons/service/snapclient/source/resources/language/English/strings.po]: https://en.wiktionary.org/wiki/autodetect so i guess `Autodetect` should be okay

File: packages/addons/addon-depends/snapcast-depends/snapcast/package.mk (hunk 1 above)
Comment (MilhouseVH) [reply packages/addons/addon-depends/snapcast-depends/snapcast/package.mk]: Yes, it's fairly common, to the point of almost being a "standard".

**Comment by awiouy** [comment]
Rebased

File: packages/addons/addon-depends/snapcast-depends/asio/package.mk [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="asio"
+PKG_VERSION="9229964"
+PKG_SHA256="799ca6607cb873bdc3253b0f8a058cb15af9c9fc2c74fd1732f6fe7726d7c200"
+PKG_ARCH="any"
+PKG_LICENSE="BSL"
```
[the remark this answers is not in the document]
Comment (awiouy) [reply packages/addons/addon-depends/snapcast-depends/asio/package.mk]: done

File: packages/addons/service/snapclient/package.mk (hunk 3 above)
Comment (awiouy) [reply packages/addons/service/snapclient/package.mk]: done

File: packages/addons/service/snapclient/source/resources/language/English/strings.po (hunk 5 above)
Comment (awiouy) [reply packages/addons/service/snapclient/source/resources/language/English/strings.po]: done

**Comment by awiouy** [comment]
I have rebased to add shaiport-sync.
I have posted preview addons [here](https://forum.libreelec.tv/thread/11817-snapcast/?postID=83859#post83859).
To test: install one Snapserver and at least one Snapclient and stream to Snapcast with airplay.
On Raspberry Pi, Snapclient can automatically activate onboard audio (set `Manage onboard audio` in addon settings).

File: packages/addons/service/snapclient/package.mk (hunk 3 above)
Comment (CvH) [reply packages/addons/service/snapclient/package.mk]: hmm still at the wrong line, should be
```
PKG_LONGDESC="..."
PKG_TOOLCHAIN="manual"

PKG_IS_ADDON="yes"
```

File: packages/addons/service/snapclient/source/settings-default.xml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<settings>
```
Comment (CvH) [inline packages/addons/service/snapclient/source/settings-default.xml]: still k17 format :)

File: packages/addons/service/snapserver/package.mk [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+PKG_SECTION="service"
+PKG_SHORTDESC="Snapserver: Synchronous multi-room audio server"
+PKG_LONGDESC="Snapclient ($PKG_VERSION) is a Snapcast server. Snapcast is a multi-room client-server audio system, where all clients are time synchronized with the server to play perfectly synced audioplays."
+
+PKG_IS_ADDON="yes"
+PKG_ADDON_NAME="Snapserver"
+PKG_ADDON_TYPE="xbmc.service"
+PKG_MAINTAINER="Anton Voyl (awiouy)"
+
+PKG_TOOLCHAIN="manual"
```
Comment (CvH) [inline packages/addons/service/snapserver/package.mk]:
```
PKG_LONGDESC="..."
PKG_TOOLCHAIN="manual"

PKG_IS_ADDON="yes"
```

File: packages/addons/service/snapserver/source/settings-default.xml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<settings>
```
Comment (CvH) [inline packages/addons/service/snapserver/source/settings-default.xml]: still k17 format

File: packages/addons/addon-depends/snapcast-depends/snapcast/package.mk [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="snapcast"
+PKG_VERSION="0.13.0"
+PKG_SHA256="8409cdb20805d945fb243324a20a24b15c418466a2026bcc6a9211a6b32fdfae"
```
Comment (MilhouseVH) [inline packages/addons/addon-depends/snapcast-depends/snapcast/package.mk]: checksum is incorrect - should be `331cfc62db8631af80f6c8798b9c359eb55e34d607af0881c3520b94f238f94e`

Comment (chewitt) [reply packages/addons/addon-depends/snapcast-depends/snapcast/package.mk]: I have this in a PR along with another couple of add-ons fixed, will post later

**Pull request merged**